src/base/static/components/molecules/add-place-button.js

```diff
@@ -68,4 +72,9 @@ const mapStateToProps = state => ({
   isGeocodeAddressBarEnabled: geocodeAddressBarEnabledSelector(state),
 });
 
-export default connect(mapStateToProps)(AddPlaceButton);
+export default connect(
+  mapStateToProps,
+  null,
+  null,
+  { forwardRef: true },
+)(AddPlaceButton);
```

Comment:
Not sure if this is the right way to handle this, but this fixes the crash. If this introduces an anti-pattern let's revisit next week; I just have to get this working again this week...

Reply:
My bad - I should have caught this when I performed the initial react-redux upgrade.

I don't think this is an anti-pattern, although I suspect that we are over-using refs. Perhaps we can revisit this when we do a re-design of the app layout. For example, we *should* be able to leverage CSS such that AddPlaceButton is positioned below ContentPanel on mobile, instead of having to track the pixel height of AddPlaceButton via a ref.

Reply:
Sounds good-- yeah, I'd rather wait until we have a better sense of the redesign until we start taking apart too many things.

FWIW the ref here (and elsewhere) isn't to position the button, it's to accurately report the rendered size of elements around the map so we can size the map container correctly. Older implementations used CSS to manage resizes, but this was prone to a race condition between CSS and Mapbox's resize method, where we could never be sure that the map was resizing after its container had been resized by CSS.

Reply:
AFAIK, we are using the place button ref to calculate the height of the button, which is used to increase ContentPanel's bottom padding in the mobile layout to make room for AddPlaceButton beneath it: https://github.com/mapseed/platform/blob/0d893b04c0bf928e2223e0ac658fde86d2c5d0db/src/base/static/components/organisms/content-panel.js#L54-L56

If we instead used a CSS layout - like flexbox - to *position* AddPlaceButton underneath ContentPanel on mobile, then we wouldn't need to calculate the height of AddPlaceButton. Maybe I'm missing something, but I don't think the AddPlaceButton ref is needed to resolve Mapbox's resizing issues.

Either way, I think we agree that we'll address this issue as part of our re-design. I just want to clarify that calculating DOM dimensions, and using refs, can be hacky - sometimes it's unavoidable, but there is usually a better way.

Reply:
We might be using it there for convenience because we're calculating it anyway, but the primary purpose for it is to calculate the map container's height, here: https://github.com/mapseed/platform/blob/master/src/base/static/utils/layout-utils.js#L102

Reply:
Ah interesting - thanks for the explanation. I forgot about those layout-utils.

I think it's possible to render the map correctly without having to calculate or hard-code the pixel heights of all those elements.

This doesn't seem like an acceptable use of refs, except for the fact that we didn't have time to refactor the existing code. Managing layout in this manner is definitely an anti-pattern, but I think we can address this when we perform a re-design.

Reply:
> I think it's possible to render the map correctly without having to calculate or hard-code the pixel heights of all those elements.

It would be great to simplify the layout system-- though what we have *is* the refactored version, refactored from the old media query-based layout system that was prone to the off-center bug.

I ran out of ideas for how to fight the off-center bug. If you can think of another approach that would be great...

Reply:
I would have to spend some time to come up with a specific solution, but I'm almost certain that React provides the tools needed to address this issue without having to resort to refs.

I've always found Discord and StackOverflow helpful when I'm not sure how to proceed with a problem: https://discordapp.com/channels/102860784329052160/103696749012467712

Reply:
I find them helpful too :). 

I searched extensively for approaches to the map resize race condition, both when we were using Leaflet and when we migrated to MapboxGL, but I couldn't find anything specific at that time. Maybe that's changed, though.

As I remember the fundamental problem is to find a way to ensure that the map container's size is invalidated after all resizing is done and rendered, whether by CSS or otherwise. Maybe there is a more React-y way to handle this...